File: app/src/main/java/com/romankaranchuk/musicplayer/presentation/ui/player/page/PlayerPageFragment.java
package com.romankaranchuk.musicplayer.presentation.ui.player.page;

import android.content.res.Configuration;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.romankaranchuk.musicplayer.R;
import com.romankaranchuk.musicplayer.data.Song;
import com.romankaranchuk.musicplayer.presentation.ui.tracklist.TrackListFragment;
import com.romankaranchuk.musicplayer.utils.MathUtils;
import com.squareup.picasso.Picasso;

public class PlayerPageFragment extends Fragment {

    private static String ARGUMENT_PAGE_NUMBER = "arg_page_number";
    private static String SAVE_PAGE_NUMBER = "save_page_number";
    private static String LOG_TAG = "myLogs";

    private ImageView albumCoverImageView;
    private TextView lyricSong;
    private int pageNumber;

    public static PlayerPageFragment newInstance(int page){
        PlayerPageFragment playerPageFragment = new PlayerPageFragment();
        Bundle arguments = new Bundle();
        arguments.putInt(ARGUMENT_PAGE_NUMBER, page);
        playerPageFragment.setArguments(arguments);
        return playerPageFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            pageNumber = getArguments().getInt(ARGUMENT_PAGE_NUMBER);
        }
        Log.d(LOG_TAG, "PlayerPageFragment onCreate: " + pageNumber);

        int savedPageNumber = -1;
        if (savedInstanceState != null){
            savedPageNumber = savedInstanceState.getInt(SAVE_PAGE_NUMBER);
        }
        Log.d(LOG_TAG, "savedPageNumber = " + savedPageNumber);
    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState){
        Log.d(LOG_TAG, "PlayerPageFragment onCreateView: " + pageNumber);
        return inflater.inflate(R.layout.fragment_player_page, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        albumCoverImageView = view.findViewById(R.id.fullscreenAlbumCover);
        lyricSong = view.findViewById(R.id.lyricSong);

        setAlbumCoverImageViewSize();

        if (pageNumber > -1 && pageNumber < TrackListFragment.getSongs().size()){
            setAlbumCoverImageView(TrackListFragment.getSongs().get(pageNumber));
            setLyricSong(TrackListFragment.getSongs().get(pageNumber));
        }
    }

    public void setAlbumCoverImageViewSize(){
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE){
            DisplayMetrics dm = new DisplayMetrics();
            getActivity().getWindowManager().getDefaultDisplay().getMetrics(dm);
            int height = dm.heightPixels ;
            getAlbumCoverImageView().setLayoutParams(new RelativeLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT, height));
        }
    }

    public ImageView getAlbumCoverImageView(){
        return albumCoverImageView;
    }

    public void setAlbumCoverImageView(Song song){
        Picasso.with(getContext()).load(song.getImagePath()).into(albumCoverImageView);
        albumCoverImageView.setTag(song.getImagePath());
        if (MathUtils.tryParse(song.getImagePath()) != -1) {
            albumCoverImageView.setScaleType(ImageView.ScaleType.CENTER);
        }
    }

    public void setLyricSong(Song song){
        lyricSong.setText(song.getLyricsSong());
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState){
        super.onSaveInstanceState(outState);
        outState.putInt(SAVE_PAGE_NUMBER, pageNumber);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState){
        super.onActivityCreated(savedInstanceState);
        Log.d(LOG_TAG, "PlayerPageFragment onActivityCreated: " + pageNumber);
    }
}

File: app/src/main/java/com/romankaranchuk/musicplayer/presentation/ui/player/page/PlayerPageFragment.kt
package com.romankaranchuk.musicplayer.presentation.ui.player.page

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import com.romankaranchuk.musicplayer.data.Song
import com.romankaranchuk.musicplayer.databinding.FragmentPlayerPageBinding
import com.romankaranchuk.musicplayer.di.util.Injectable
import com.romankaranchuk.musicplayer.utils.MathUtils
import com.squareup.picasso.Picasso
import timber.log.Timber
import javax.inject.Inject

class PlayerPageFragment : Fragment(), Injectable {

    private var _binding: FragmentPlayerPageBinding? = null
    private val binding get() = _binding!!

    @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
    private val viewModel: PlayerPageViewModel by viewModels { viewModelFactory }

    private var pageNumber = 0

    override fun onAttach(context: Context) {
        super.onAttach(context)

        lifecycle.addObserver(viewModel)
    }

    override fun onDetach() {
        super.onDetach()

        lifecycle.removeObserver(viewModel)
    }

    override fun onDestroy() {
        super.onDestroy()

        _binding = null
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        Timber.d("onCreateView: $pageNumber")
        _binding = FragmentPlayerPageBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (arguments != null) {
            pageNumber = arguments!!.getInt(ARGUMENT_PAGE_NUMBER)
        }
        Timber.d("onCreate: $pageNumber")
        var savedPageNumber = -1
        if (savedInstanceState != null) {
            savedPageNumber = savedInstanceState.getInt(SAVE_PAGE_NUMBER)
        }
        Timber.d("savedPageNumber = $savedPageNumber")

        setAlbumCoverImageViewSize()

        bindViewModels()
        viewModel.onCreateCalled(pageNumber)
    }

    private fun bindViewModels() {
        viewModel.state.observe(viewLifecycleOwner) {
            when (it) {
                is PlayerPageViewModel.State.onCreate -> {
                    setAlbumCoverImageView(it.song)
                }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(SAVE_PAGE_NUMBER, pageNumber)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        Timber.d("onActivityCreated: $pageNumber")
    }

    fun setAlbumCoverImageViewSize() {
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            val dm = DisplayMetrics()
            activity!!.windowManager.defaultDisplay.getMetrics(dm)
            val height = dm.heightPixels
            binding.fullscreenAlbumCover.layoutParams = RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, height)
        }
    }

    fun setAlbumCoverImageView(song: Song) {
        Picasso.get().load(song.imagePath).into(binding.fullscreenAlbumCover)
        binding.fullscreenAlbumCover.tag = song.imagePath
        if (MathUtils.tryParse(song.imagePath) != -1) {
            binding.fullscreenAlbumCover.scaleType = ImageView.ScaleType.CENTER
        }
    }

    companion object {
        private const val ARGUMENT_PAGE_NUMBER = "arg_page_number"
        private const val SAVE_PAGE_NUMBER = "save_page_number"

        fun newInstance(page: Int): PlayerPageFragment {
            return PlayerPageFragment().apply {
                arguments = Bundle().apply {
                    putInt(ARGUMENT_PAGE_NUMBER, page)
                }
            }
        }
    }
}